src/librustc_mir/borrow_check/nll/constraint_generation.rs

```diff
@@ -75,14 +84,19 @@ impl<'cx, 'gcx, 'tcx> ConstraintGeneration<'cx, 'gcx, 'tcx> {
                     }
                 });
 
-            self.liveness
-                .drop
-                .simulate_block(self.mir, bb, |location, live_locals| {
+            self.liveness.drop.simulate_block(
```

Comment:
Before we call `simulate_block`, we need to do:

```
self.flow_inits.reset_to_entry_of(bb);
```

This will reset the state to the entry of the basic block. Heh, I just realized a kind of annoying problem. The `flow_inits` wants to walk **forward**, whereas liveness wants to walk **backwards**. This means we can't readily "update" the flow-in-progress for each step as we go.

I think then that we what we ought to do is to walk the liveness backwards and collect the live variables into a vector. Something like this:

```rust
let mut live_locals: Vec<(Location, Vec<Local>)> = vec![];
self.liveness.drop.simulate_block(self.mir, bb, |location, live_locals| {
    live_locals.push((location, live_locals.iter().collect()));
});

self.flow_inits.reset_to_entry_of(bb);
for (index, statement) in block_data.statements.iter() {
    let location = Location { block: bb, statement: index };
    let (location2, live_locals) = live_locals.pop().unwrap();
    assert_eq!(location, location2);

    // code you had before:
    let mpi = self.move_data.rev_lookup.find_local(live_local);
    if self.flow_inits.has_any_child_of(mpi).is_some() {
        let live_local_ty = self.mir.local_decls[live_local].ty;
        self.add_drop_live_constraint(live_local_ty, location);
    }

    self.flow_inits.reconstruct_statement_effect(location, statement);
}

// same as above, but for terminator, probably can attach some kind of helper function
```